### PR TITLE
CLI Tag Functionality 

### LINKED ISSUE #2141: [CLI] Support tag management (create, list, delete)
Requested in https://github.com/huggingface/huggingface_hub/issues/960#issuecomment-2013533386.

Similar to https://github.com/huggingface/huggingface_hub/issues/1655 (branch management from CLI)

### PR DESCRIPTION
**Pull request by bilgehanertan**
Resolves #2141 

Since I wasn't sure whether to add unit tests about functionalities directly, I have only added parsing tests. Depending on the reviews, I can add tests using the token from constants. 

I was also unsure how to handle the token argument `tag.py:120`; please verify whether this is the proper way to do so. 

@Wauplin

### CODE REVIEW

File: src/huggingface_hub/commands/tag.py
Comment (Wauplin) [inline src/huggingface_hub/commands/tag.py:27]: Hi @bilgehanertan, thanks for the PR! I haven't thoroughly reviewed it but implementation-wise the structure looks good. Before moving forward, I'd like to suggest some changes in the CLI format. Here is how I'd envision it:

```sh
# Tag model
$ huggingface-cli tag username/my-model 1.0
# Tag model with message
$ huggingface-cli tag username/my-model 1.0 -m "First release"
$ huggingface-cli tag username/my-model 1.0 --message "First release"
# Tag model with revision
$ huggingface-cli tag username/my-model 1.0 --revision develop
# Tag dataset
$ huggingface-cli tag username/my-dataset 1.0 --repo-type dataset

# List all model tags
$ huggingface-cli tag -l username/my-model
$ huggingface-cli tag --list username/my-model
# List all dataset tags
$ huggingface-cli tag -l username/my-dataset --repo-type dataset

# Delete model tag
$ huggingface-cli tag -d username/my-model 1.0
$ huggingface-cli tag --delete username/my-model 1.0
``` 

Main changes are:
- removed `create`/`delete`/`list` commands in favor of `-l`/`--list` and `-d`/`--delete` to list and delete tags. By default, tag creation is expected.
- updated `--type` to `--repo-type` for consistency
- I think namespace of the repo should always be explicit. 
- updated `--tag-message` to `--message` or `-m`

With this format, we would be much closer of the [`git tag`](https://git-scm.com/docs/git-tag)  command line that is already well established. What do you think?

Comment (bilgehanertan) [reply src/huggingface_hub/commands/tag.py:27]: @Wauplin, thanks for the feedback! 
I have adjusted it & tested it; it should look better now.

File: src/huggingface_hub/commands/tag.py
Comment (Wauplin) [inline src/huggingface_hub/commands/tag.py:180]:
```suggestion
            if choice not in ("", "y", "yes"):
```

(slightly easier to read)

**Review by Wauplin — COMMENTED** [review]
Thanks for the changes @bilgehanertan! API looks good now. Left some comment to simplify a bit + review some messages but overall looks very promising :) Would you mind doing a last iteration on it? Thanks in advance!

**Comment by bilgehanertan** [comment]
> Thanks for the changes @bilgehanertan! API looks good now. Left some comment to simplify a bit + review some messages but overall looks very promising :) Would you mind doing a last iteration on it? Thanks in advance!

@Wauplin all done!

File: src/huggingface_hub/commands/tag.py
Comment (Wauplin) [inline src/huggingface_hub/commands/tag.py:101]:
```suggestion
```

(no need for confirmation when creating a tag IMO (not destructive action))

**Review by Wauplin — COMMENTED** [review]
Thanks @bilgehanertan! I played with it locally and it's :fire:!! I pushed a few tweaks for consistency with the CLI. PR looks good to me logic-wise now. Before merging, would you mind documenting this new command on this page: https://huggingface.co/docs/huggingface_hub/guides/cli?

You'll need to update [this file](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/guides/cli.md?plain=1) to document `huggingface-cli tag`. Section can go between `huggingface-cli delete-cache` and `huggingface-cli env`. Usually in the doc guide with put "real examples". If you are looking for repos with tags on the Hub, on can cite the [bigcode/the-stack](https://huggingface.co/datasets/bigcode/the-stack) dataset or the [Wauplin/gradio-space-ci](https://huggingface.co/spaces/Wauplin/gradio-space-ci) Space. Thanks in advance!

**Comment by bilgehanertan** [comment]
> Thanks @bilgehanertan! I played with it locally and it's 🔥!! I pushed a few tweaks for consistency with the CLI. PR looks good to me logic-wise now. Before merging, would you mind documenting this new command on this page: https://huggingface.co/docs/huggingface_hub/guides/cli?
> 
> You'll need to update [this file](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/guides/cli.md?plain=1) to document `huggingface-cli tag`. Section can go between `huggingface-cli delete-cache` and `huggingface-cli env`. Usually in the doc guide with put "real examples". If you are looking for repos with tags on the Hub, on can cite the [bigcode/the-stack](https://huggingface.co/datasets/bigcode/the-stack) dataset or the [Wauplin/gradio-space-ci](https://huggingface.co/spaces/Wauplin/gradio-space-ci) Space. Thanks in advance!

@Wauplin I have added the necessary guides, hope it looks good!

File: docs/source/en/guides/cli.md
Comment (Wauplin) [inline docs/source/en/guides/cli.md:461]:
```suggestion
To tag a repo, you need to provide the `repo_id` and the `tag` name:
```

File: docs/source/en/guides/cli.md
Comment (Wauplin) [inline docs/source/en/guides/cli.md:459]:
```suggestion
### Tag a model
```

File: docs/source/en/guides/cli.md
Comment (Wauplin) [inline docs/source/en/guides/cli.md:469]:
```suggestion
### Tag a model at a specific revision
```

File: docs/source/en/guides/cli.md
Comment (Wauplin) [inline docs/source/en/guides/cli.md:471]:
```suggestion
If you want to tag a specific revision, you can use the `--revision` option. By default, the tag will be created on the `main` branch:
```

File: docs/source/en/guides/cli.md
Comment (Wauplin) [inline docs/source/en/guides/cli.md:480]:
```suggestion
### Tag a dataset or a Space

If you want to tag a dataset or Space, you must specify the `--repo-type` option:
```

File: docs/source/en/guides/cli.md
Comment (Wauplin) [inline docs/source/en/guides/cli.md:513]:
```suggestion

You can also pass `-y` to skip the confirmation step.

```

**Review by Wauplin — APPROVED** [review]
Nice! That looks great! Thanks a lot for the reactivity :heart:

**Review by Wauplin — APPROVED** [review]
PR's clean and is ready to go I think! Let's wait for the CI to pass and I'll merge. Nice API, nice docs, nice tests, thanks a lot for your contribution and iterations @bilgehanertan! :heart:

**Pull request merged**
